### PR TITLE
Refactor cluster templates

### LINKED ISSUE #855: Reduce the code duplication in the cluster templates
/kind feature
/area provider/ibmcloud

**Describe the solution you'd like**
Most of the code [here](https://github.com/kubernetes-sigs/cluster-api-provider-ibmcloud/tree/main/templates) is duplicate, lets explore a way to avoid this.

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

### PR DESCRIPTION
**Pull request by Amulyam24**
**What this PR does / why we need it**:
This PR refactors the construction of cluster templates by reducing code duplication.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #855 

**Special notes for your reviewer**:

/area provider/ibmcloud

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Release note**:

```release-note
Refactor cluster templates
```

### CODE REVIEW

**Comment by Prajyot-Parab** [comment]
@Amulyam24 
Is this the same behavior on your machine as well?

```
cluster-api-provider-ibmcloud % make generate-templates
./hack/version.sh: line 95: GIT_MAJOR: unbound variable
hack/tools/bin/kustomize build /Users/prajyotparab/Workspace/sandbox/cluster-api-provider-ibmcloud/templates/cluster-template --load-restrictor LoadRestrictionsNone > /Users/prajyotparab/Workspace/sandbox/cluster-api-provider-ibmcloud/templates/cluster-template.yaml
# Warning: 'patchesStrategicMerge' is deprecated. Please use 'patches' instead. Run 'kustomize edit fix' to update your Kustomization automatically.
hack/tools/bin/kustomize build /Users/prajyotparab/Workspace/sandbox/cluster-api-provider-ibmcloud/templates/cluster-template-powervs --load-restrictor LoadRestrictionsNone > /Users/prajyotparab/Workspace/sandbox/cluster-api-provider-ibmcloud/templates/cluster-template-powervs.yaml
# Warning: 'patchesStrategicMerge' is deprecated. Please use 'patches' instead. Run 'kustomize edit fix' to update your Kustomization automatically.
hack/tools/bin/kustomize build /Users/prajyotparab/Workspace/sandbox/cluster-api-provider-ibmcloud/templates/cluster-template-powervs-cloud-provider --load-restrictor LoadRestrictionsNone > /Users/prajyotparab/Workspace/sandbox/cluster-api-provider-ibmcloud/templates/cluster-template-powervs-cloud-provider.yaml
# Warning: 'patchesStrategicMerge' is deprecated. Please use 'patches' instead. Run 'kustomize edit fix' to update your Kustomization automatically.
hack/tools/bin/kustomize build /Users/prajyotparab/Workspace/sandbox/cluster-api-provider-ibmcloud/templates/cluster-template-powervs-clusterclass --load-restrictor LoadRestrictionsNone > /Users/prajyotparab/Workspace/sandbox/cluster-api-provider-ibmcloud/templates/cluster-template-powervs-clusterclass.yaml
hack/tools/bin/kustomize build /Users/prajyotparab/Workspace/sandbox/cluster-api-provider-ibmcloud/templates/cluster-template-vpc-load-balancer --load-restrictor LoadRestrictionsNone > /Users/prajyotparab/Workspace/sandbox/cluster-api-provider-ibmcloud/templates/cluster-template-vpc-load-balancer.yaml
# Warning: 'patchesStrategicMerge' is deprecated. Please use 'patches' instead. Run 'kustomize edit fix' to update your Kustomization automatically.
```

**Comment by Amulyam24** [comment]
> @Amulyam24 Is this the same behavior on your machine as well?
> 
> ```
> cluster-api-provider-ibmcloud % make generate-templates
> ./hack/version.sh: line 95: GIT_MAJOR: unbound variable
> hack/tools/bin/kustomize build /Users/prajyotparab/Workspace/sandbox/cluster-api-provider-ibmcloud/templates/cluster-template --load-restrictor LoadRestrictionsNone > /Users/prajyotparab/Workspace/sandbox/cluster-api-provider-ibmcloud/templates/cluster-template.yaml
> # Warning: 'patchesStrategicMerge' is deprecated. Please use 'patches' instead. Run 'kustomize edit fix' to update your Kustomization automatically.
> hack/tools/bin/kustomize build /Users/prajyotparab/Workspace/sandbox/cluster-api-provider-ibmcloud/templates/cluster-template-powervs --load-restrictor LoadRestrictionsNone > /Users/prajyotparab/Workspace/sandbox/cluster-api-provider-ibmcloud/templates/cluster-template-powervs.yaml
> # Warning: 'patchesStrategicMerge' is deprecated. Please use 'patches' instead. Run 'kustomize edit fix' to update your Kustomization automatically.
> hack/tools/bin/kustomize build /Users/prajyotparab/Workspace/sandbox/cluster-api-provider-ibmcloud/templates/cluster-template-powervs-cloud-provider --load-restrictor LoadRestrictionsNone > /Users/prajyotparab/Workspace/sandbox/cluster-api-provider-ibmcloud/templates/cluster-template-powervs-cloud-provider.yaml
> # Warning: 'patchesStrategicMerge' is deprecated. Please use 'patches' instead. Run 'kustomize edit fix' to update your Kustomization automatically.
> hack/tools/bin/kustomize build /Users/prajyotparab/Workspace/sandbox/cluster-api-provider-ibmcloud/templates/cluster-template-powervs-clusterclass --load-restrictor LoadRestrictionsNone > /Users/prajyotparab/Workspace/sandbox/cluster-api-provider-ibmcloud/templates/cluster-template-powervs-clusterclass.yaml
> hack/tools/bin/kustomize build /Users/prajyotparab/Workspace/sandbox/cluster-api-provider-ibmcloud/templates/cluster-template-vpc-load-balancer --load-restrictor LoadRestrictionsNone > /Users/prajyotparab/Workspace/sandbox/cluster-api-provider-ibmcloud/templates/cluster-template-vpc-load-balancer.yaml
> # Warning: 'patchesStrategicMerge' is deprecated. Please use 'patches' instead. Run 'kustomize edit fix' to update your Kustomization automatically.
> ```

yes @Prajyot-Parab. Made changes for the warning regarding deprecation of `patchesStrategicMerge` except for `cluster-template-vpc-load-balancer` since it has multiple patches in one, will see if that can be changed too.

**Comment by Prajyot-Parab** [comment]
> @Amulyam24 Is this the same behavior on your machine as well?

> yes @Prajyot-Parab. Made changes for the warning regarding deprecation of `patchesStrategicMerge` except for `cluster-template-vpc-load-balancer` since it has multiple patches in one, will see if that can be changed too.

@Amulyam24 Checked looking good, let me know if its possible fix `vpc-load-blancer` template as well.

**Comment by Amulyam24** [comment]
> > @Amulyam24 Is this the same behavior on your machine as well?
> 
> > yes @Prajyot-Parab. Made changes for the warning regarding deprecation of `patchesStrategicMerge` except for `cluster-template-vpc-load-balancer` since it has multiple patches in one, will see if that can be changed too.
> 
> @Amulyam24 Checked looking good, let me know if its possible to fix `vpc-load-blancer` template as well.

@Prajyot-Parab, split the patches. PR is ready for review. PTAL

**Review by mkumatag — APPROVED** [review]
/lgtm

File: Makefile [hunk 1]
```diff
@@ -174,6 +175,14 @@ generate-go-conversions: $(CONVERSION_GEN) ## Generate conversions go code
 		--output-file-base=zz_generated.conversion $(CONVERSION_GEN_OUTPUT_BASE) \
 		--go-header-file=./hack/boilerplate/boilerplate.generatego.txt
 
+.PHONY: generate-templates
```
Comment (mkumatag) [inline Makefile:178]: We should explore how consume this target! @Amulyam24 can you create a follow up task for that?

**Review by mkumatag — CHANGES_REQUESTED** [review]

**Comment by mkumatag** [comment]
lets merge this meanwhile

/approve

File: templates/cluster-template-vpc-load-balancer.yaml [hunk 2]
```diff
@@ -406,4 +403,7 @@ data:
                 name: ibm-cloud-config
             - name: ibm-secret
               secret:
-                secretName: ibm-cloud-credential
+                secretName: ibm-cloud-credentia
```
Comment (Prajyot-Parab) [inline templates/cluster-template-vpc-load-balancer.yaml:406]: `ibm-cloud-credentia -> ibm-cloud-credential`

@Amulyam24 typo error

File: templates/addons/crs.yaml [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+                requests:
+                  cpu: 200m
+          hostNetwork: true
+          volumes:
+            - name: ibm-config-volume
+              configMap:
+                name: ibm-cloud-config
+            - name: ibm-secret
+              secret:
+                secretName: ibm-cloud-credentia
```
Comment (Prajyot-Parab) [inline templates/addons/crs.yaml:256]: `ibm-cloud-credentia -> ibm-cloud-credential`

@Amulyam24 another one here

**Pull request merged**